src/radical/pilot/agent/launch_method/mpirun.py

```diff
@@ -15,114 +15,176 @@ class MPIRun(LaunchMethod):
+            lm_info['omplace'] = ru.which('omplace')
+            lm_info['mpt']     = True
 
-        # do we need ccmrun or dplace?
-        if '_ccmrun' in self.name:
-            self._ccmrun = ru.which('ccmrun')
-            if not self._ccmrun:
-                raise RuntimeError("ccmrun not found!")
+        mpi_version, mpi_flavor = self._get_mpi_info(lm_info['command'])
+        lm_info['mpi_version']  = mpi_version
+        lm_info['mpi_flavor']   = mpi_flavor
 
-        if '_dplace' in self.name:
-            self._dplace = ru.which('dplace')
-            if not self._dplace:
-                raise RuntimeError("dplace not found!")
+        return lm_info
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _init_from_info(self, lm_info, lm_cfg):
+
+        self._env         = lm_info['env']
+        self._env_sh      = lm_info['env_sh']
+        self._command     = lm_info['command']
+
+        self._mpt         = lm_info['mpt']
+        self._rsh         = lm_info['rsh']
+        self._ccmrun      = lm_info['ccmrun']
+        self._dplace      = lm_info['dplace']
+        self._omplace     = lm_info['omplace']
 
-        self.mpi_version, self.mpi_flavor = \
-                                       self._get_mpi_info(self.launch_command)
+        self._mpi_version = lm_info['mpi_version']
+        self._mpi_flavor  = lm_info['mpi_flavor']
 
 
     # --------------------------------------------------------------------------
     #
-    def construct_command(self, t, launch_script_hop):
+    def finalize(self):
 
-        slots        = t['slots']
-        uid          = t['uid']
-        td          = t['description']
-        sandbox      = t['task_sandbox_path']
-        task_exec    = td['executable']
-        task_threads = td.get('cpu_threads', 1)
-        task_env     = td.get('environment') or dict()
-        task_args    = td.get('arguments')   or list()
-        task_argstr  = self._create_arg_string(task_args)
+        pass
 
-        if '_dplace' in self.name and task_threads > 1:
-            # dplace pinning would disallow threads to map to other cores
-            raise ValueError('dplace can not place threads [%d]' % task_threads)
 
-        # Construct the executable and arguments
-        if task_argstr: task_command = "%s %s" % (task_exec, task_argstr)
-        else          : task_command = task_exec
+    # --------------------------------------------------------------------------
+    #
+    def can_launch(self, task):
+
+        # mpirun can launch any executable
+        if task['description']['executable']:
+            return True
+        return False
+
+
+    # --------------------------------------------------------------------------
+    #
+    def get_launcher_env(self):
+
+        return ['. $RP_PILOT_SANDBOX/%s' % self._env_sh]
+
+
+    # --------------------------------------------------------------------------
+    #
+    def get_task_env(self, spec):
```

Comment:
this should be moved to LM base class? (currently only 3 LMs have it, and it is duplicated for all of them)

Reply:
Hmm - they are identical, true - but also they need to remove the launcher env, and in principle that is launcher specific.  Well, I can move it, let's see how it goes...